src/lib.rs

```diff
@@ -437,7 +441,13 @@ impl RawIter {
 
     fn size_hint<T: Send>(&self, thread_local: &ThreadLocal<T>) -> (usize, Option<usize>) {
         let total = thread_local.values.load(Ordering::Acquire);
-        (total - self.yielded, None)
+        // max calculation: we know that all non-null buckets may be filled and as every
+        // bucket can contain 2^bucket_id items, which means that all current buckets
+        // combined have n^max_bucket - 1 items
```

Comment:
That's not actually correct since new elements can be added concurrently while we are iterating. As such we can't make any promises about an upper bound.

Reply:
From the documentation of `Iterator::size_hint`:
    /// # Implementation notes
    ///
    /// It is not enforced that an iterator implementation yields the declared
    /// number of elements. A buggy iterator may yield less than the lower bound
    /// or more than the upper bound of elements.
    ///
    /// `size_hint()` is primarily intended to be used for optimizations such as
    /// reserving space for the elements of the iterator, but must not be
    /// trusted to e.g., omit bounds checks in unsafe code. An incorrect
    /// implementation of `size_hint()` should not lead to memory safety
    /// violations.
    ///
    /// That said, the implementation should provide a correct estimation,
    /// because otherwise it would be a violation of the trait's protocol.

Reply:
From these notes it seems that a good estimate of the upper bound could lead to performance improvements even though it might not be exact.

Reply:
The documentation specifically says that such an iterator is buggy, it's just that unsafe code should be reliable when passed buggy impls. `Vec` uses the lower bound for reserving memory anyways, so this shouldn't be a big problem.

Reply:
Fixed